go-controller/pkg/ovn/controller/egress_services/egress_services_controller.go

```diff
@@ -417,16 +490,337 @@ func (c *Controller) repair() error {
+		return
+	}
+	c.egressServiceQueue.Add(key)
+}
+
+// onEgressServiceUpdate queues the EgressService for processing.
+func (c *Controller) onEgressServiceUpdate(oldObj, newObj interface{}) {
+	oldEQ := oldObj.(*egressserviceapi.EgressService)
+	newEQ := newObj.(*egressserviceapi.EgressService)
+
+	if oldEQ.ResourceVersion == newEQ.ResourceVersion ||
+		!newEQ.GetDeletionTimestamp().IsZero() {
+		return
+	}
+
+	key, err := cache.MetaNamespaceKeyFunc(newObj)
+	if err == nil {
+		c.egressServiceQueue.Add(key)
+	}
+}
+
+// onEgressServiceDelete queues the EgressService for processing.
+func (c *Controller) onEgressServiceDelete(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+	c.egressServiceQueue.Add(key)
+}
+
+func (c *Controller) runEgressServiceWorker(wg *sync.WaitGroup) {
+	for c.processNextEgressServiceWorkItem(wg) {
+	}
+}
+
+func (c *Controller) processNextEgressServiceWorkItem(wg *sync.WaitGroup) bool {
+	wg.Add(1)
+	defer wg.Done()
+
+	key, quit := c.egressServiceQueue.Get()
+	if quit {
+		return false
+	}
+
+	defer c.egressServiceQueue.Done(key)
+
+	err := c.syncEgressService(key.(string))
+	if err == nil {
+		c.egressServiceQueue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
+
+	if c.egressServiceQueue.NumRequeues(key) < maxRetries {
+		c.egressServiceQueue.AddRateLimited(key)
+		return true
+	}
+
+	c.egressServiceQueue.Forget(key)
+	return true
+}
+
+func (c *Controller) syncEgressService(key string) error {
+	c.Lock()
+	defer c.Unlock()
+
+	startTime := time.Now()
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	klog.Infof("Processing sync for Egress Service %s/%s", namespace, name)
+
+	defer func() {
+		klog.V(4).Infof("Finished syncing Egress Service %s/%s : %v", namespace, name, time.Since(startTime))
+	}()
+
+	es, err := c.egressServiceLister.EgressServices(namespace).Get(name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	svc, err := c.serviceLister.Services(namespace).Get(name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	state := c.services[key]
+	if es == nil {
+		if state == nil {
+			// The EgressService was deleted and wasn't configured before.
+			// We delete it from the unallocated service cache just in case.
+			delete(c.unallocatedServices, key)
+			return nil
+		}
+		// The service is configured but does no longer have the egress service resource,
+		// meaning we should clear all of its resources.
+		return c.clearServiceResources(key, state)
+	}
+
+	if svc == nil {
+		if state == nil {
+			// The service object was deleted and was not an allocated egress service.
+			// We delete it from the unallocated service cache just in case.
+			delete(c.unallocatedServices, key)
+			return c.setEgressServiceHost(namespace, name, "")
```

Comment:
EgressSVC sets the status in two places, in the CR and as a label on a node.
I think we might want to check if `es.Status.Host` is set and clear the label from the node if it is.

Reply:
Doesn't the same general reasoning also apply to the EgressSVC status field then?
I am cool with addressing the issue separately but considering that we store the same status (selected node) in two places I would be surprised if we need to handle those in different ways.

Reply:
what I meant is while the controller can keep the status field synced to its cache all the time, it can't (currently) make sure the nodes are synced.
i.e if the user manually sets the status it would be set back to the correct node by the controller (same if the user removes the label from the node)
thinking about it more maybe the label issue can be tackled from the node controller, but I rather addressing it later :smile: